File: cube.cpp
#include "cube.h"
#include "map.h"
#include <string>
#include "colors.h"

Cube::Cube() {
	level = 0;
}

Cube::Cube(int level) {
	loadCube(level);
}

Map* Cube::getFront() {
	return &sides[0];
}

Map* Cube::getTop() {
	return &sides[1];
}

Map* Cube::getLeft() {
	return &sides[2];
}

Map* Cube::getRight() {
	return &sides[3];
}

Map* Cube::getBottom() {
	return &sides[4];
}

Map* Cube::getBack() {
	return &sides[5];
}

int Cube::getLevel() {
	return level;
}

/*
*	Rotates the cube up by changing the sides, and rotating them as needed.
*/
void Cube::rotateUp() {

	getBack()->xAxisFlip();
	
	Map temp = sides[0];
	sides[0] = sides[4];
	sides[4] = sides[5];
	sides[5] = sides[1];
	sides[1] = temp;
	
	
	getRight()->rotateClockwise();

	getLeft()->rotateCounterClockwise();

	getBack()->xAxisFlip();
	
}

/*
*	Rotates the cube down by changing the sides, and rotating them as needed.
*/
void Cube::rotateDown() {
	getBack()->xAxisFlip();
	
	Map temp = sides[0];
	sides[0] = sides[1];
	sides[1] = sides[5];
	sides[5] = sides[4];
	sides[4] = temp;

	getLeft()->rotateClockwise();
	getRight()->rotateCounterClockwise();

	getBack()->xAxisFlip();
}

/*
*	Rotates the cube left by changing the sides, and rotating them as needed.
*/
void Cube::rotateLeft() {

	//invert y axis of the back
	getBack()->yAxisFlip();

	Map temp = sides[0];
	sides[0] = sides[3];
	sides[3] = sides[5];
	sides[5] = sides[2];
	sides[2] = temp;

	getTop()->rotateClockwise();
	getBottom()->rotateCounterClockwise();

	//invert y axis of the back
	getBack()->yAxisFlip();
}

/*
*	Rotates the cube right by changing the sides, and rotating them as needed.
*/
void Cube::rotateRight() {

	//invert y axis of the back
	getBack()->yAxisFlip();

	Map temp = sides[0];
	sides[0] = sides[2];
	sides[2] = sides[5];
	sides[5] = sides[3];
	sides[3] = temp;

	getTop()->rotateCounterClockwise();
	getBottom()->rotateClockwise();

	//invert y axis of the back
	getBack()->yAxisFlip();
}


Map* Cube::getSides() {
	return sides;
}

/*
*
*/
bool Cube::loadCube(int newLevel){
	bool isSuccess = true;
	char str[36];

	level = newLevel;

	memset(str,'\0',36);
	sprintf(str,"levels\\level_%i\\front.txt", level);
	sides[0].setRotation(0);
	sides[0].setColor(Colors::RED);
	isSuccess = sides[0].loadMap(str);

	memset(str,'\0',36);
	sprintf(str,"levels\\level_%i\\top.txt", level);
	sides[1].setRotation(0);
	sides[1].setColor(Colors::BLUE);
	isSuccess = sides[1].loadMap(str);

	memset(str,'\0',36);
	sprintf(str,"levels\\level_%i\\left.txt", level);
	sides[2].setRotation(0);
	sides[2].setColor(Colors::GREEN);
	isSuccess = sides[2].loadMap(str);

	memset(str,'\0',36);
	sprintf(str,"levels\\level_%i\\right.txt", level);
	sides[3].setRotation(0);
	sides[3].setColor(Colors::YELLOW);
	isSuccess = sides[3].loadMap(str);

	memset(str,'\0',36);
	sprintf(str,"levels\\level_%i\\bottom.txt", level);
	sides[4].setRotation(0);
	sides[4].setColor(Colors::WHITE);
	isSuccess = sides[4].loadMap(str);

	memset(str,'\0',36);
	sprintf(str,"levels\\level_%i\\back.txt", level);
	sides[5].setRotation(0);
	sides[5].setColor(Colors::PURPLE);
	isSuccess = sides[5].loadMap(str);

	return isSuccess;
} 

bool Cube::reloadCube() {
	return loadCube(level);
}

bool Cube::loadNextCube() {
	return loadCube(level+1);
}

File: assets/cube.js
import Map from './map.js';

const FACE_KEYS = ['front', 'top', 'left', 'right', 'bottom', 'back'];

export default class Cube {
  constructor(levels = []) {
    this.levels = levels;
    this.levelId = 0;
    this.sides = Array.from({ length: FACE_KEYS.length }, () => new Map());
  }

  getFront() {
    return this.sides[0];
  }

  getTop() {
    return this.sides[1];
  }

  getLeft() {
    return this.sides[2];
  }

  getRight() {
    return this.sides[3];
  }

  getBottom() {
    return this.sides[4];
  }

  getBack() {
    return this.sides[5];
  }

  getSides() {
    return this.sides;
  }

  getLevel() {
    return this.levelId;
  }

  rotateUp() {
    this.getBack().xAxisFlip();

    const temp = this.sides[0];
    this.sides[0] = this.sides[4];
    this.sides[4] = this.sides[5];
    this.sides[5] = this.sides[1];
    this.sides[1] = temp;

    this.getRight().rotateClockwise();
    this.getLeft().rotateCounterClockwise();

    this.getBack().xAxisFlip();
  }

  rotateDown() {
    this.getBack().xAxisFlip();

    const temp = this.sides[0];
    this.sides[0] = this.sides[1];
    this.sides[1] = this.sides[5];
    this.sides[5] = this.sides[4];
    this.sides[4] = temp;

    this.getLeft().rotateClockwise();
    this.getRight().rotateCounterClockwise();

    this.getBack().xAxisFlip();
  }

  rotateLeft() {
    this.getBack().yAxisFlip();

    const temp = this.sides[0];
    this.sides[0] = this.sides[3];
    this.sides[3] = this.sides[5];
    this.sides[5] = this.sides[2];
    this.sides[2] = temp;

    this.getTop().rotateClockwise();
    this.getBottom().rotateCounterClockwise();

    this.getBack().yAxisFlip();
  }

  rotateRight() {
    this.getBack().yAxisFlip();

    const temp = this.sides[0];
    this.sides[0] = this.sides[2];
    this.sides[2] = this.sides[5];
    this.sides[5] = this.sides[3];
    this.sides[3] = temp;

    this.getTop().rotateCounterClockwise();
    this.getBottom().rotateClockwise();

    this.getBack().yAxisFlip();
  }

  loadCube(levelId) {
    const levelData = this.levels.find((lvl) => lvl.id === levelId);
    if (!levelData) {
      return false;
    }
    this.levelId = levelId;

    const { faces } = levelData;
    this.sides[0] = Map.fromLines(faces.front, 'front');
    this.sides[1] = Map.fromLines(faces.top, 'top');
    this.sides[2] = Map.fromLines(faces.left, 'left');
    this.sides[3] = Map.fromLines(faces.right, 'right');
    this.sides[4] = Map.fromLines(faces.bottom, 'bottom');
    this.sides[5] = Map.fromLines(faces.back, 'back');

    this.sides.forEach((side) => side.setRotation(0));
    return true;
  }

  reloadCube() {
    return this.loadCube(this.levelId);
  }

  loadNextCube() {
    return this.loadCube(this.levelId + 1);
  }
}
